src/Servers/HttpSys/src/RequestProcessing/ResponseBody.cs

```diff
@@ -188,30 +192,41 @@ private unsafe void FlushInternal(bool endOfRequest, ArraySegment<byte> data = n
         }
     }
 
-    private List<GCHandle> PinDataBuffers(bool endOfRequest, ArraySegment<byte> data, out HttpApiTypes.HTTP_DATA_CHUNK[] dataChunks)
+    private unsafe Span<GCHandle> BuildDataChunks(scoped ref UnmanagedBufferAllocator allocator, bool endOfRequest, ArraySegment<byte> data, out Span<HttpApiTypes.HTTP_DATA_CHUNK> dataChunks)
     {
-        var pins = new List<GCHandle>();
+        const int maxGCHandleCount = 4;
+        int gcHandleIndex = 0;
+        var pins = allocator.AllocAsSpan<GCHandle>(maxGCHandleCount);
+
+        // Manually initialize the allocated GCHandles
+        for (int i = 0; i < maxGCHandleCount; ++i)
```

Comment:
```suggestion
        for (int i = 0; i < pins.Length; ++i)
```
This will allow bounds-check elimination when accessing the Span to kick in.

Reply:
Alternatively, you can do `pins.Clear()`.